### DISCUSSION TITLE
Does Fresh Framework have a feature like Next.js's getStaticPaths and generateStaticParams to statically generate routes at build time?

### DISCUSSION THREAD

**Discussion by m7medVision** [freshframework/fresh#2267] [Q&A]
I've structured my Fresh app to include a [slug].tsx file that renders .md files for blog-like content. However, I think this not efficient way, as the server seems to re-render the same content for each request.
Does Fresh Framework have a feature like Next.js's getStaticPaths and generateStaticParams to statically generate routes at build time?

**Comment by CAYdenberg** [comment]
No, because the Fresh-y way of doing things is just-in-time rendering rather than having a build step. 

So, the way to respond dynamically to routes with parameterized URLs is to have a route like `[slug].tsx` and then figure out what to render in the handler. If for example the slug does not exist you can then call `context.renderNotFound()` and defer to your 404 route.

If you give me an example of what you would do with getStaticPaths I can give you a more detailed explanation.

**Reply by m7medVision** [reply]
I've structured my Fresh app to include a `[slug].tsx` file that renders `.md` files for blog-like content. However, I think this not efficient way, as the server seems to re-render the same content for each request. 
I appreciate your help, @CAYdenberg .

**Reply by CAYdenberg** [reply]
So in that sense Fresh is similar to Rails or Wordpress in that it has to go through the work of rendering on each request. In your case this would also include retrieving the .md file over the filesystem. The next level above that would be an in-memory cache, but before you do that make sure you understand:
- that you aren't prematurely optimizing ... does this _seem_ like the wrong approach because of your familiarity with Next, or is your app _actually_ slow
- where is the slowness coming from? My instinct is the filesystem, but it could also be parsing and/or rendering markdown, or something else.

**Reply by m7medVision** [reply]
Hi @CAYdenberg, thanks so much for taking the time to explain this to me. I'm still new to Fresh and I want to make sure that my app performs well. I was hoping to find a way to avoid unnecessary re-rendering, and I understand that caching might be good in terms of scaling for the future. My app is performing well now, but I'm concerned about scalability as the website grows.

**Reply by CAYdenberg** [reply]
No problem at all. 

FWIW, I find Fresh _feels_ faster than other systems I've worked with, because it's engineered to deliver _far_ fewer assets. It all depends on what parts of performance you care about ... but in my not so humble opinion, network latency is a bigger contributor to user frustration than either slow servers or inefficient frontend code. 

But they are all different kinds of performance and require entirely different approaches.
